exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillAggregateExpandGroupingSetsRule.java

```diff
@@ -0,0 +1,426 @@
+      // will be evaluated later using the $g column
+      Aggregate newAggregate;
+      if (aggregate instanceof DrillAggregateRel) {
+        newAggregate = new DrillAggregateRel(
+            cluster,
+            aggregate.getTraitSet(),
+            input,
+            groupSet,
+            ImmutableList.of(groupSet),
+            regularAggCalls);
+      } else {
+        newAggregate = aggregate.copy(
+            aggregate.getTraitSet(),
+            input,
+            groupSet,
+            ImmutableList.of(groupSet),
+            regularAggCalls);
+      }
+
+      // Create a project to add NULLs for missing grouping columns
+      List<RexNode> projects = new ArrayList<>();
+      List<String> fieldNames = new ArrayList<>();
+
+      // Add grouping columns (with NULLs for columns not in this grouping set)
+      int aggOutputIdx = 0;
+      int outputColIdx = 0; // Index in the final output row type
+      for (int col : fullGroupSet) {
+        if (groupSet.get(col)) {
+          // Column is in this grouping set - project it directly from the aggregate output
+          RexNode inputRef = rexBuilder.makeInputRef(newAggregate, aggOutputIdx);
+          projects.add(inputRef);
+          aggOutputIdx++;
+        } else {
+          // Column is NOT in this grouping set - project a typed NULL literal
+          // Use the expected output type from the original aggregate to create a properly typed NULL
+          // This prevents type inference issues in the UNION ALL
+          org.apache.calcite.rel.type.RelDataType nullType =
+              aggregate.getRowType().getFieldList().get(outputColIdx).getType();
+          // Use makeLiteral with null value and explicit type to create a typed NULL
+          projects.add(rexBuilder.makeNullLiteral(nullType));
+        }
+        fieldNames.add(aggregate.getRowType().getFieldList().get(outputColIdx).getName());
+        outputColIdx++;
+      }
+
+      // Add aggregate result columns (only regular aggregates, not GROUPING functions)
+      // We'll use the alias from the original aggregate call
+      for (int j = 0; j < regularAggCalls.size(); j++) {
+        projects.add(rexBuilder.makeInputRef(newAggregate, aggOutputIdx));
+        AggregateCall regCall = regularAggCalls.get(j);
+        String fieldName = regCall.getName() != null ? regCall.getName() : ("$f" + (fullGroupSet.cardinality() + j));
+        fieldNames.add(fieldName);
+        aggOutputIdx++;
+      }
+
+      // Add grouping ID column ($g)
+      // The grouping ID is a bitmap where bit i is 1 if column i is NOT in the grouping set
+      int groupingId = 0;
+      int bitPosition = 0;
+      for (int col : fullGroupSet) {
+        if (!groupSet.get(col)) {
+          groupingId |= (1 << bitPosition);
+        }
+        bitPosition++;
+      }
+      projects.add(rexBuilder.makeLiteral(groupingId, typeFactory.createSqlType(org.apache.calcite.sql.type.SqlTypeName.INTEGER), true));
+      fieldNames.add("$g");
```

Comment:
I would prefer extracting `$g` into a named constant (e.g., `GROUPING_ID_COLUMN_NAME`) to improve readability and eliminate magic literals in the codebase.

Reply:
Fixed